src/app/api/models/webcal/webcal.service.ts

```diff
@@ -0,0 +1,62 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+import { Webcal } from './webcal';
+import { EntityService } from '../entity.service';
+
+interface WebcalUpdateParams {
+  enabled?: boolean;
+  should_change_guid?: boolean;
+  include_start_dates?: boolean;
+}
+
+@Injectable()
+export class WebcalService extends EntityService<Webcal> {
+
+  entityName: string = 'Webcal';
+  protected readonly endpointFormat = 'webcal';
+
+  constructor(
+    httpClient: HttpClient
+  ) {
+    super(httpClient);
+  }
+
+  protected createInstanceFrom(json: any, other?: any): Webcal {
+
+    // If the API returns null (in the case the user does not have a webcal), return immediately.
+    if (!json) return null;
+
+    // Otherwise, create Webcal instance. =
+    let webcal = new Webcal();
+    webcal.updateFromJson(json);
+    return webcal;
+  }
+
+  public keyForJson(json: any): string {
+    return json.id;
+  }
+
+  /**
+   * Performs an HTTP PUT request against the /webcal endpoint, with the specified data.
+   * `EntityService.put` does not work here because the endpoint returns `null` when the user's webcal is disabled.
```

Comment:
Is the endpoint our own, or some webcal library on the backend? Can we change the backend to return something other than null when no cal exists?

Reply:
Hiya Jake 👋
Since this feature requires the user to manually opt-in to it, the `Webcal` entity on the backend currently represents the availability of the web calendar for a user, i.e. the user `has_one` webcal if they have enabled it, or has none if its disabled (in which case the API returns `null`).

https://github.com/doubtfire-lms/doubtfire-api/blob/cc13c5630e5f486c75b61b385ed1f2cc7392f40f/app/api/webcal_api.rb#L18

I could wrap this response in a hash that'll change the format to,

```json
{
  "enabled": true,
  "webcal": {
    ...
  }
}
```

Thoughts?

Reply:
Yes, something like that would be better. Then we can remove this custom implementation of the EntityService PUT, and it's more clear.